File: t-tasks/src/main/java/com/teo/ttasks/data/local/TaskFields.java
package com.teo.ttasks.data.local;

import android.support.annotation.Nullable;

import com.teo.ttasks.data.model.TTask;

import java.util.Date;
import java.util.HashMap;

import static com.teo.ttasks.data.model.Task.STATUS_COMPLETED;
import static com.teo.ttasks.data.model.Task.STATUS_NEEDS_ACTION;

public class TaskFields extends HashMap<String, Object> {

    private static final String KEY_TITLE = "title";
    private static final String KEY_DUE = "due";
    private static final String KEY_NOTES = "notes";
    private static final String KEY_COMPLETED = "completed";
    private static final String KEY_STATUS = "status";

    public static TaskFields fromTask(TTask task) {
        final TaskFields taskFields = new TaskFields();
        taskFields.putTitle(task.getTitle());
        taskFields.putCompleted(task.isCompleted(), task.getCompleted());
        taskFields.putDueDate(task.getDue());
        taskFields.putNotes(task.getNotes());
        return taskFields;
    }

    public void putTitle(@Nullable String title) {
        if (title == null || title.isEmpty()) {
            remove(KEY_TITLE);
            return;
        }
        put(KEY_TITLE, title);
    }

    public void putDueDate(@Nullable Date dueDate) {
        if (dueDate == null) {
            remove(KEY_DUE);
            return;
        }
        put(KEY_DUE, dueDate);
    }

    public void putNotes(@Nullable String notes) {
        put(KEY_NOTES, notes);
    }

    public void putCompleted(boolean isCompleted, @Nullable Date completed) {
        if (isCompleted) {
            put(KEY_STATUS, STATUS_COMPLETED);
        } else {
            put(KEY_STATUS, STATUS_NEEDS_ACTION);
        }
        put(KEY_COMPLETED, completed);
    }

    public String getTitle() {
        return (String) get(KEY_TITLE);
    }

    public Date getDueDate() {
        return (Date) get(KEY_DUE);
    }

    public String getNotes() {
        return (String) get(KEY_NOTES);
    }
}

File: t-tasks/src/main/java/com/teo/ttasks/data/local/TaskFields.kt
package com.teo.ttasks.data.local

import com.evernote.android.job.util.support.PersistableBundleCompat
import com.teo.ttasks.data.model.Task
import java.util.*

private const val NUM_FIELDS = 5

// TODO: remove this, it's unnecessary
class TaskFields : HashMap<String, String?>(NUM_FIELDS) {

    companion object {
        private const val KEY_TITLE = "title"
        private const val KEY_DUE = "due"
        private const val KEY_NOTES = "notes"
        private const val KEY_COMPLETED = "completed"
        private const val KEY_STATUS = "status"

        fun taskFields(block: TaskFields.() -> Unit): TaskFields = TaskFields().apply(block)

        /**
         * Convert bundle date to TaskFields. This is used when de-serializing task data
         * in a job
         */
        fun fromBundle(bundle: PersistableBundleCompat): TaskFields? {
            val taskFields = taskFields {
                if (bundle.containsKey(KEY_TITLE)) {
                    title = bundle[KEY_TITLE] as String
                }
                if (bundle.containsKey(KEY_DUE)) {
                    dueDate = bundle[KEY_DUE] as String
                }
                if (bundle.containsKey(KEY_NOTES)) {
                    notes = bundle[KEY_NOTES] as String
                }
                if (bundle.containsKey(KEY_COMPLETED)) {
                    completed = bundle[KEY_COMPLETED] as String?
                }
                if (bundle.containsKey(KEY_STATUS)) {
                    this[KEY_STATUS] = bundle[KEY_STATUS] as String
                }
            }
            return if (taskFields.isNotEmpty()) taskFields else null
        }
    }

    var title
        get() = this[KEY_TITLE]
        set(title) {
            when {
                // Disallow tasks with empty or no title
                title == null || title.isBlank() -> remove(KEY_TITLE)
                else -> this[KEY_TITLE] = title
            }
        }

    /**
     * The task's due date, always in UTC.
     */
    var dueDate
        get() = this[KEY_DUE]
        set(dueDate) {
            when (dueDate) {
                null -> remove(KEY_DUE)
                else -> this[KEY_DUE] = dueDate
            }
        }

    var notes
        get() = this[KEY_NOTES]
        set(notes) {
            when {
                notes == null || notes.isBlank() -> remove(KEY_NOTES)
                else -> this[KEY_NOTES] = notes
            }
        }

    var completed
        get() = this[KEY_COMPLETED]
        set(completed) {
            this[KEY_COMPLETED] = completed
            this[KEY_STATUS] = if (completed == null) Task.STATUS_NEEDS_ACTION else Task.STATUS_COMPLETED
        }
}
